src/fklearn/training/pipeline.py

```diff
@@ -7,10 +8,33 @@
 from fklearn.types import LearnerFnType, LearnerReturnType, PredictFnType
 
 
-def build_pipeline(*learners: LearnerFnType) -> LearnerFnType:
+def _has_one_unfilled_arg(learner: LearnerFnType) -> None:
+    no_default_list = [p for p, a in signature(learner).parameters.items() if a.default == '__no__default__']
+    if len(no_default_list) > 1:
+        raise ValueError("Learner {0} has more than one unfilled argument: {1}\n"
+                         "Make sure all learners are curried properly and only require one argument,"
+                         " which is the dataset (usually `df`)."
+                         .format(learner.__name__, ', '.join(no_default_list)))
+
+
+def _no_variable_args(learner: LearnerFnType, predict_fn: PredictFnType) -> None:
+    invalid_parameter_kinds = (Parameter.VAR_POSITIONAL, Parameter.VAR_KEYWORD)
+    var_args = [p for p, a in signature(predict_fn).parameters.items() if a.kind in invalid_parameter_kinds]
+    if len(var_args) != 0:
+        raise ValueError("Predict function of learner {0} contains variable length arguments: {1}\n"
+                         "Make sure no predict function uses arguments like *args or **kwargs."
+                         .format(learner.__name__, ', '.join(var_args)))
+
+
+def _check_unfilled_arg(*learners: LearnerFnType) -> None:
+    for l in learners:
+        _has_one_unfilled_arg(l)
+
+
+def build_pipeline_repeated_learners(*learners: LearnerFnType) -> LearnerFnType:
```

Comment:
This doesn't look like the best approach, as we may need to change the same code in two places. From what I've saw we only change two lines, the one with the `defaultdict`, and the one where we append the logs. I think we can rewrite this in a better way to avoid code duplication.

I think you can add a parameter to the `build_pipeline` method that defaults to False, something like `has_repeated_learners`, and them use this to change the way we initialise the serialisation dict, and the way we append data. Could be something like:

```python
def build_pipeline(*learners: LearnerFnType, has_duplicated_learners: bool = False) -> LearnerFnType:
(....)
serialisation = defaultdict(list) if has_duplicated_learners else {}  # type: dict
(....)
if has_duplicated_learners:
    serialisation[learner_name].append({"fn": learner_fn, "log": learner_log, **model_objects})
else
    serialisation[learner_name] = {"fn": learner_fn, "log": learner_log, **model_objects}
```

This approach isn't a breaking change, and we avoid a lot of code duplication

WDYT?

Reply:
Great idea! I just refactored the code and asked for your review again.